File: Programers/Level2/타겟넘버.py
def dfs(numbers, idx, total, target):
    global answer
    if idx == len(numbers):
        # print(total)
        if total == target:
            answer += 1
        return
            
    dfs(numbers, idx+1, total + numbers[idx], target)
    dfs(numbers, idx+1, total - numbers[idx], target)
    
    
answer = 0
def solution(numbers, target):
    
    dfs(numbers, 0, 0, target)
    return answer
    
# def solution(numbers, target):
#     global answer
#     total = 0
#     dfs(numbers, target, total, 0)
    
#     return answer

# answer = 0
# def dfs(arr, target, total, idx):
#     global answer
#     if idx == len(arr):
#         if total == target:
#             answer += 1
#         return

#     total += arr[idx]
#     dfs(arr, target, total, idx+1)
#     total -= arr[idx]
#     total -= arr[idx]
#     dfs(arr,target, total, idx+1)
    
    
    

File: Javascript/Programmers/타겟넘버.js
function solution(numbers, target) {
    let answer = 0;
    let arr = numbers;
    let total = 0;

    function dfs(arr, total, target, idx) {
        if (idx === arr.length) {
            if (total === target) {
                answer++;
            }
            return;
        }

        dfs(arr, total + arr[idx], target, idx + 1);
        dfs(arr, total - arr[idx], target, idx + 1);
    }

    dfs(arr, total, target, 0);
    return answer;
}
